### ISSUE TITLE
[8.0] CryptographicException : An unknown chain building error occurred - Failing in test System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback

### ISSUE THREAD

**Issue by carlossanlop** [dotnet/runtime#99425]
## Build Information
Build: https://dev.azure.com/dnceng-public/public/_build/results?buildId=593479
Build error leg or test failing: System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback

## Error Message

Fill the error message using [step by step known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssues.md#how-to-fill-out-a-known-issue-error-section).

```json
{
  "ErrorMessage": "System.Security.Cryptography.CryptographicException : An unknown chain building error occurred",
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```

- Affected PR: https://github.com/dotnet/runtime/pull/99407
- Queue: `Libraries Test Run release coreclr windows x64 Debug`
- Job result: &view=logs&j=cb131f87-d5af-5ecc-6e5f-6dc86994859c&t=76503e0a-8c41-542c-43da-63a4758a95b0&s=6884a131-87da-5381-61f3-d7acc3b91d76
- Log file: https://helixre107v0xd1eu3ibi6ka.blob.core.windows.net/dotnet-runtime-refs-pull-99407-merge-463928b535334e2bb2/System.Net.Security.Tests/3/console.6e5c5769.log?helixlogtype=result
- Output:
```
===========================================================================================================

C:\h\w\9CD108B8\w\BFEF0A95\e>"C:\h\w\9CD108B8\p\dotnet.exe" exec --runtimeconfig System.Net.Security.Tests.runtimeconfig.json --depsfile System.Net.Security.Tests.deps.json xunit.console.dll System.Net.Security.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing  
  Discovering: System.Net.Security.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Net.Security.Tests (found 429 of 504 test cases)
  Starting:    System.Net.Security.Tests (parallel test collections = on, max threads = 2)
    System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback(checkRevocation: True) [FAIL]
      System.Security.Cryptography.CryptographicException : An unknown chain building error occurred.
      Stack Trace:
        /_/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509Chain.cs(192,0): at System.Security.Cryptography.X509Certificates.X509Chain.Build(X509Certificate2 certificate, Boolean throwOnException)
        /_/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509Chain.cs(91,0): at System.Security.Cryptography.X509Certificates.X509Chain.Build(X509Certificate2 certificate)
        /_/src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.cs(63,0): at System.Net.Security.SslStreamCertificateContext.Create(X509Certificate2 target, X509Certificate2Collection additionalCertificates, Boolean offline, SslCertificateTrust trust, Boolean noOcspFetch)
        /_/src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.cs(32,0): at System.Net.Security.SslStreamCertificateContext.Create(X509Certificate2 target, X509Certificate2Collection additionalCertificates, Boolean offline, SslCertificateTrust trust)
        /_/src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.cs(27,0): at System.Net.Security.SslStreamCertificateContext.Create(X509Certificate2 target, X509Certificate2Collection additionalCertificates, Boolean offline)
        /_/src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs(262,0): at System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback_Core(X509RevocationMode revocationMode, Nullable`1 offlineContext, Boolean noIntermediates)
        /_/src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs(310,0): at System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback_Core(X509RevocationMode revocationMode, Nullable`1 offlineContext, Boolean noIntermediates)
        /_/src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs(310,0): at System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback_Core(X509RevocationMode revocationMode, Nullable`1 offlineContext, Boolean noIntermediates)
        /_/src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs(310,0): at System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback_Core(X509RevocationMode revocationMode, Nullable`1 offlineContext, Boolean noIntermediates)
        /_/src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs(310,0): at System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback_Core(X509RevocationMode revocationMode, Nullable`1 offlineContext, Boolean noIntermediates)
        --- End of stack trace from previous location ---
    System.Net.Security.Tests.NegotiatedCipherSuiteTest.CipherSuitesPolicy_OnlyTls13CipherSuiteAllowedButChosenProtocolsDoesNotAllowIt_Fails [SKIP]
      Condition(s) not met: "CipherSuitesPolicyAndTls13Supported"
    System.Net.Security.Tests.NegotiatedCipherSuiteTest.CipherSuitesPolicy_OnlyNonTls13CipherSuiteAllowedButOtherSideDoesNotAllowIt_Fails [SKIP]
      Condition(s) not met: "CipherSuitesPolicySupported"
    System.Net.Security.Tests.NegotiatedCipherSuiteTest.CipherSuitesPolicy_AllowSameTwoOnBothSidesLessPreferredIsTls13_Success [SKIP]
      Condition(s) not met: "CipherSuitesPolicySupported"
    System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback(checkRevocation: False) [FAIL]
      System.Security.Cryptography.CryptographicException : An unknown chain building error occurred.
      Stack Trace:
        /_/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509Chain.cs(192,0): at System.Security.Cryptography.X509Certificates.X509Chain.Build(X509Certificate2 certificate, Boolean throwOnException)
        /_/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509Chain.cs(91,0): at System.Security.Cryptography.X509Certificates.X509Chain.Build(X509Certificate2 certificate)
        /_/src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.cs(63,0): at System.Net.Security.SslStreamCertificateContext.Create(X509Certificate2 target, X509Certificate2Collection additionalCertificates, Boolean offline, SslCertificateTrust trust, Boolean noOcspFetch)
        /_/src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.cs(32,0): at System.Net.Security.SslStreamCertificateContext.Create(X509Certificate2 target, X509Certificate2Collection additionalCertificates, Boolean offline, SslCertificateTrust trust)
        /_/src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.cs(27,0): at System.Net.Security.SslStreamCertificateContext.Create(X509Certificate2 target, X509Certificate2Collection additionalCertificates, Boolean offline)
        /_/src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs(262,0): at System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback_Core(X509RevocationMode revocationMode, Nullable`1 offlineContext, Boolean noIntermediates)
        /_/src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs(310,0): at System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback_Core(X509RevocationMode revocationMode, Nullable`1 offlineContext, Boolean noIntermediates)
        /_/src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs(310,0): at System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback_Core(X509RevocationMode revocationMode, Nullable`1 offlineContext, Boolean noIntermediates)
        /_/src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs(310,0): at System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback_Core(X509RevocationMode revocationMode, Nullable`1 offlineContext, Boolean noIntermediates)
        /_/src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs(310,0): at System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback_Core(X509RevocationMode revocationMode, Nullable`1 offlineContext, Boolean noIntermediates)
        --- End of stack trace from previous location ---
```

 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=593479
**Error message validated:** `[System.Security.Cryptography.CryptographicException : An unknown chain building error occurred`]
**Result validation:** :white_check_mark: Known issue matched with the provided build.
**Validation performed at:** 3/7/2024 7:12:44 PM UTC

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[609327](https://dev.azure.com/dnceng-public/public/_build/results?buildId=609327)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=609327&view=ms.vss-test-web.build-test-results-tab&runId=14866956&resultId=166170)||
|[609324](https://dev.azure.com/dnceng-public/public/_build/results?buildId=609324)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=609324&view=ms.vss-test-web.build-test-results-tab&runId=14866698&resultId=162719)||
|[606476](https://dev.azure.com/dnceng-public/public/_build/results?buildId=606476)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=606476&view=ms.vss-test-web.build-test-results-tab&runId=14795308&resultId=136947)|dotnet/runtime#99449|
|[605539](https://dev.azure.com/dnceng-public/public/_build/results?buildId=605539)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=605539&view=ms.vss-test-web.build-test-results-tab&runId=14740352&resultId=158244)||
|[604791](https://dev.azure.com/dnceng-public/public/_build/results?buildId=604791)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=604791&view=ms.vss-test-web.build-test-results-tab&runId=14718946&resultId=166074)|dotnet/runtime#99696|
|[602571](https://dev.azure.com/dnceng-public/public/_build/results?buildId=602571)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=602571&view=ms.vss-test-web.build-test-results-tab&runId=14651720&resultId=139420)||
|[602738](https://dev.azure.com/dnceng-public/public/_build/results?buildId=602738)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=602738&view=ms.vss-test-web.build-test-results-tab&runId=14649606&resultId=143424)||
|[602568](https://dev.azure.com/dnceng-public/public/_build/results?buildId=602568)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=602568&view=ms.vss-test-web.build-test-results-tab&runId=14639050&resultId=171543)||
|[602274](https://dev.azure.com/dnceng-public/public/_build/results?buildId=602274)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=602274&view=ms.vss-test-web.build-test-results-tab&runId=14635304&resultId=158965)|dotnet/runtime#99586|
|[602023](https://dev.azure.com/dnceng-public/public/_build/results?buildId=602023)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=602023&view=ms.vss-test-web.build-test-results-tab&runId=14623964&resultId=174066)||
|[601852](https://dev.azure.com/dnceng-public/public/_build/results?buildId=601852)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=601852&view=ms.vss-test-web.build-test-results-tab&runId=14618372&resultId=159322)|dotnet/runtime#99651|
|[601886](https://dev.azure.com/dnceng-public/public/_build/results?buildId=601886)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=601886&view=ms.vss-test-web.build-test-results-tab&runId=14617330&resultId=200870)||
|[600824](https://dev.azure.com/dnceng-public/public/_build/results?buildId=600824)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=600824&view=ms.vss-test-web.build-test-results-tab&runId=14572842&resultId=147471)||
|[599827](https://dev.azure.com/dnceng-public/public/_build/results?buildId=599827)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=599827&view=ms.vss-test-web.build-test-results-tab&runId=14545642&resultId=162939)|dotnet/runtime#99607|
|[599209](https://dev.azure.com/dnceng-public/public/_build/results?buildId=599209)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=599209&view=ms.vss-test-web.build-test-results-tab&runId=14527500&resultId=140803)||
|[599197](https://dev.azure.com/dnceng-public/public/_build/results?buildId=599197)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=599197&view=ms.vss-test-web.build-test-results-tab&runId=14524322&resultId=170095)||
|[599189](https://dev.azure.com/dnceng-public/public/_build/results?buildId=599189)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=599189&view=ms.vss-test-web.build-test-results-tab&runId=14523302&resultId=165702)|dotnet/runtime#99583|
|[598852](https://dev.azure.com/dnceng-public/public/_build/results?buildId=598852)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=598852&view=ms.vss-test-web.build-test-results-tab&runId=14516002&resultId=147091)||
|[598019](https://dev.azure.com/dnceng-public/public/_build/results?buildId=598019)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=598019&view=ms.vss-test-web.build-test-results-tab&runId=14491740&resultId=180301)|dotnet/runtime#99473|
|[593199](https://dev.azure.com/dnceng-public/public/_build/results?buildId=593199)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=593199&view=ms.vss-test-web.build-test-results-tab&runId=14359852&resultId=140706)||
|[593479](https://dev.azure.com/dnceng-public/public/_build/results?buildId=593479)|dotnet/runtime|[System.Net.Security.Tests.CertificateValidationRemoteServer.ConnectWithRevocation_WithCallback](https://dev.azure.com/dnceng-public/public/_build/results?buildId=593479&view=ms.vss-test-web.build-test-results-tab&runId=14350342&resultId=163339)|dotnet/runtime#99407|
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|2|12|21|

**Comment by rzikm** [comment]
I am able to reproduce the failure on my machine, even on main, but I can't figure out what is wrong. The failure seems to originate from

https://github.com/dotnet/runtime/blob/125b73eb94738e415f05d4213c0ae25498462e0a/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/ChainPal.Windows.BuildChain.cs#L68

and the Last error is 0x80090006, which is apparently Invalid Signature. @bartonjs, would you know what could be happening? We are generating the test certificates using the `CertificateAuthority.BuildPrivatePki` helper, and later call to SslStreamCertificateContext.Create fails.

https://github.com/dotnet/runtime/blob/524598aecf36bc74f51df7af9a63448b7ec7f63d/src/libraries/System.Net.Security/tests/FunctionalTests/CertificateValidationRemoteServer.cs#L198-L265

**Comment by bartonjs** [comment]
> @bartonjs, would you know what could be happening?

Nope :(.  I haven't seen that coming out of the X509 tests, and they use that same helper a lot.  I'm not sure if the invalid signature error is relevant to the overall failure, or just happens to be the last thing sitting in the error field.

My best instinct is something wonky like one of the inputs to the chain build getting parallel disposed.

**Comment by rzikm** [comment]
I tracked it down, the failure is triggered when, *... drum roll ...* the `testName` parameter to `BuildPrivatePki` is empty, which leads to funny Subjects like `CN=server.example, O=""` on the end entity cert

**Comment by rzikm** [comment]
Fixed in main and 8.0
